utilities/hydrofabric_test/test_ngen.py

```diff
@@ -0,0 +1,198 @@
+    """
+
+    Args:
+        hydrofabric (_type_): _description_
+    """
+    attributes = gpd.read_file(hydrofabric, layer="flowpath_attributes").set_index('id')
+    x_walk = pd.Series( attributes[ ~ attributes['rl_gages'].isna() ]['rl_gages'] )
+    
+    data = {}
+    for wb, gage in x_walk.items():
+        data[wb] = {'Gage_no':[gage]}
+    import json
+    with open('crosswalk.json', 'w') as fp:
+        json.dump(data, fp, indent=2)
+
+def make_geojson(hydrofabric: Path):
+    """Create the various required geojson/json files from the geopkg
+
+    Args:
+        hydrofabric (Path): path to hydrofabric geopkg
+    """
+    try:
+        catchments = gpd.read_file(hydrofabric, layer="divides")
+        nexuses = gpd.read_file(hydrofabric, layer="nexus")
+        flowpaths = gpd.read_file(hydrofabric, layer="flowpaths")
+        edge_list = pd.DataFrame(gpd.read_file(hydrofabric, layer="flowpath_edge_list").drop(columns='geometry'))
+        make_x_walk(hydrofabric)
+        catchments.to_file("catchments.geojson")
+        nexuses.to_file("nexus.geojson")
+        flowpaths.to_file("flowpaths.geojson")
+        edge_list.to_json("flowpath_edge_list.json", orient='records', indent=2)
+    except Exception as e:
+        print(f"Unable to use hydrofabric file {hydrofabric}")
+        print(str(e))
+        raise NgenRunException
+
+def make_realization(workdir: Path, config: Formulation, forcing: Forcing, time: Time, include_routing: bool=False):
+    """Generates a ngen realization file based on the user defined formulation.
+
+    Args:
+        workdir (Path): Location to stage the realization file in.
+        config (Formulation): Formulation to build.
+        forcing (Forcing): Forcing configuration to use.
+        time (Time): Simulation time range.
+        include_routing (bool, optional): Whether or not to add a routing block to the realization. Defaults to False.
+                                          When true, assumes the routing config is in the workdir and called test_routing_config.yaml
+    """
+
+    f = Formulation(name=config['name'], params=config)
+    r = Realization(formulations=[f], forcing=forcing)
+    if(include_routing):
+        routing = {'t_route_config_file_with_path': workdir/'test_routing_config.yaml'}
+    else:
+        routing = None
+    g = NgenRealization(global_config=r, time=time, routing=routing)
+    with open(workdir/"realization.json", 'w') as fp:
+         fp.write( g.json(by_alias=True, exclude_none=True, indent=4))
+
+#TODO on failue of check_call, output log to console???
+def test_ngen(hydrofabric: Union[str, Path], config, tag=""):
+    """_summary_
+
+    Args:
+        hydrofabric (Union[str, Path]): path to geopackage hydrofabric to test
+        config (General): the user configuration
+        tag (str, optional): A tag to prefix the temporary directory the model runs in. Defaults to "".
+
+    Raises:
+        NgenRunException: _description_
+    """
+    ngen = config.ngen
```

Comment:
We may want to resolve the path to this if its not in `PATH`. If it is not in `PATH` and you dont specify it as an absolute path, I don't think this will work.